experimental/packages/web-common/test/SessionManager.test.ts

```diff
@@ -0,0 +1,184 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as assert from 'assert';
+import { SessionManager, SessionManagerConfig } from '../src/SessionManager';
+import { SessionIdGenerator } from '../src/types/SessionIdGenerator';
+import { SessionStore } from '../src/types/SessionStore';
+import { SessionObserver } from '../src/types/SessionObserver';
+import { Session } from '../src/types/Session';
+
+class MockSessionIdGenerator implements SessionIdGenerator {
+  private _id = 0;
+  generateSessionId(): string {
+    return `session-${++this._id}`;
+  }
+}
+
+class MockSessionStore implements SessionStore {
+  private _session: Session | null = null;
+  save(session: Session): void {
+    this._session = session;
+  }
+  get(): Session | null {
+    return this._session;
+  }
+}
+
+class MockSessionObserver implements SessionObserver {
+  startedSessions: Session[] = [];
+  endedSessions: Session[] = [];
+
+  onSessionStarted(newSession: Session, oldSession?: Session): void {
+    this.startedSessions.push(newSession);
+  }
+
+  onSessionEnded(session: Session): void {
+    this.endedSessions.push(session);
+  }
+}
+
+describe('SessionManager', () => {
+  let config: SessionManagerConfig;
+  let store: MockSessionStore;
+  let idGenerator: MockSessionIdGenerator;
+  let observer: MockSessionObserver;
+  let sessionManager: SessionManager;
+
+  beforeEach(() => {
+    idGenerator = new MockSessionIdGenerator();
+    store = new MockSessionStore();
+    observer = new MockSessionObserver();
+    config = {
+      sessionIdGenerator: idGenerator,
+      sessionStore: store,
+    };
+  });
+
+  afterEach(() => {
+    sessionManager.shutdown();
+  });
+
+  it('should start a new session if none exists', () => {
+    sessionManager = new SessionManager(config);
+    const session = sessionManager.getSession();
+    assert.strictEqual(session?.id, 'session-1');
+    assert.strictEqual(store.get()?.id, 'session-1');
+  });
+
+  it('should return the same session ID if session exists', () => {
+    sessionManager = new SessionManager(config);
+    sessionManager.getSessionId(); // Starts session-1
+    const sessionId = sessionManager.getSessionId(); // Reuse session-1
+    assert.strictEqual(sessionId, 'session-1');
+  });
+
+  it('should reset session after max duration', async () => {
+    config = {
+      sessionIdGenerator: idGenerator,
+      sessionStore: store,
+      maxDuration: 0.5,
+    };
+    sessionManager = new SessionManager(config);
+
+    sessionManager.getSessionId(); // Starts session-1
+    await wait(600);
```

Comment:
Hmm, would it be possible to use sinon fake timers instead to avoid depending on actual timing for these tests? 🤔